python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,505 @@
+
+    def _generate_number_samples_weights(self):
+        """
+        Function generates an array `samples` of ints of samples and their
+        weights that can be used for generating X and dataset.
+        """
+        samples = np.random.choice(np.arange(self.randind,
+                                             self.randind + 3),
+                                   self.nsamples_random)
+        maxsample = np.max(samples)
+        w = np.empty(self.nsamples_random, dtype=np.float32)
+        for i in range(self.nsamples_exact, self.nsamples_random):
+            w[i] = shapley_kernel(self.M, samples[i])
+        samples = cp.array(samples, dtype=np.int32)
+        w = cp.array(w)
+        return samples, w
+
+    def _weighted_linear_regression(self, y_hat, nonzero_inds=None):
+        if nonzero_inds is None:
+            y_hat = y_hat - self.expected_value
+
+            Aw = self.mask * cp.sqrt(self.weights[:, cp.newaxis])
+            Bw = y_hat * cp.sqrt(self.weights)
+            X, *_ = cp.linalg.lstsq(Aw, Bw)
+
+            return X
+        else:
+            y_hat = y_hat[nonzero_inds] - self.expected_value
+            Aw = self.mask[nonzero_inds] * cp.sqrt(
+                self.weights[nonzero_inds, cp.newaxis]
+            )
+            Bw = y_hat * cp.sqrt(self.weights[nonzero_inds])
+            X, *_ = cp.linalg.lstsq(Aw, Bw)
+
+            return X
+
+    def shap_values(self, X, l1_reg='auto'):
+        """
+        Legacy interface to estimate the SHAP values for a set of samples.
+
+        Parameters
+        ----------
+        X : Dense matrix containing floats or doubles.
+            Acceptable formats: CUDA array interface compliant objects like
+            CuPy, cuDF DataFrame/Series, NumPy ndarray and Pandas
+            DataFrame/Series.
+        l1_reg : str (default: 'auto')
+            The l1 regularization to use for feature selection.
+
+        Returns
+        -------
+        array or list
+
+        """
+        return self.explain(X, l1_reg)
+
+    def __call__(self,
+                 X,
+                 l1_reg='auto'):
+        if has_shap():
+            warn("SHAP's Explanation object is still experimental, the main "
+                 "API currently is 'explainer.shap_values'.")
+            from shap import Explanation
+            res = self.explain(X, l1_reg)
+            out = Explanation(
+                values=res,
+                base_values=self.expected_value,
+                base_values=self.expected_value,
+                data=self.background,
+                feature_names=self.feature_names,
+            )
+            return out
+        else:
+            raise ImportError("SHAP package required to build Explanation "
+                              "object. Use the explainer.shap_values "
+                              "function to get the shap values, or install "
+                              "SHAP to use the new API style.")
+
+
+@lru_cache(maxsize=None)
```

Comment:
Are you sure we want the `maxsize` to be None here? Possibly we can call `binomCoef.cache_clear()` after every call to `explain()`?

Reply:
We could bound them somewhat, but the thing is that for example `_binomCoef` _at most_ will have `n` entries in the cache, but `n` is a runtime value so we cannot use it as the `maxsize` parameter. So leaving it at `None` looks like the appropriate course in this case, given what we know the maximum number of cached values (and each entry is a single number, so memory consumption will be small even for large `n`, say 500,000)